### DISCUSSION TITLE
Material homogenization for 3D printed parts (FDM)

### DISCUSSION THREAD

**Discussion by yvanblanchard** [adtzlr/felupe#838] [Q&A]
Hello
I would like to know how to embed FDM printing imperfections (voids between filaments and layers) in the material using homogenization models (ex: average ratio of voids = modified stiffness values).
Thank you

**Comment by adtzlr** [comment]
Hi @yvanblanchard,

do you mean how to assign different material parameters per cell / quadrature point? The array of a material parameter must be of shape `(nquadraturepoints, ncells)` of the region on which the solid body is defined. This works for strain energy density functions like [`NeoHooke`](https://felupe.readthedocs.io/en/latest/felupe/constitution/core.html#felupe.NeoHooke) and is also supported in AD-based strain energy function classes like [`Hyperelastic`](https://felupe.readthedocs.io/en/latest/felupe/constitution/hyperelasticity.html#felupe.Hyperelastic).

```python
import felupe as fem
import numpy as np

mesh = fem.Cube(n=6)
region = fem.RegionHexahedron(mesh)
field = fem.FieldContainer([fem.Field(region, dim=3)])
boundaries, loadcase = fem.dof.uniaxial(field, clamped=True)

# material parameter field must have broadcast-compatible shape with
# (nquadraturepoints, ncells)
mu = 1.0 + np.random.rand(1, region.mesh.ncells)
umat = fem.NeoHooke(mu=mu, bulk=2)
solid = fem.SolidBody(umat=umat, field=field)

move = fem.math.linsteps([0, 1], num=5)
ramp = {boundaries["move"]: move}
step = fem.Step(items=[solid], ramp=ramp, boundaries=boundaries)

job = fem.Job(steps=[step])
job.evaluate()

solid.plot("Principal Values of Cauchy Stress").show()
```

![grafik](https://github.com/user-attachments/assets/d2caabd8-5b2a-4686-a68f-b03566d22009)

**Comment by yvanblanchard** [comment]
My question was much more related to how micro-scale imperfections (because of FDM printing process) could be applied at marco-scale ,in Felupe (through material model).
In particular, voids lead to lower mechnical performances, and one way to handle it is too adapt the elastic modulus of material from void ratio.

Voids (FDM):
![image](https://github.com/user-attachments/assets/8daff2ff-24d0-40e2-a300-7dc30cf09b71)

Relationship between voids ratio and elastic modulus:
![image](https://github.com/user-attachments/assets/e60c0687-c66c-43b7-aef1-3fd440719926)

It could be useful to have this kind of material model settings, for more accurate structural analysis of 3D printed parts.
But maybe it is out of the scope of the library !...

**Reply by adtzlr** [reply]
Thanks! I thought you meant some geometric variation of the elastic properties.

It's very interesting but at least it is out of scope from my personal knowledge 😃. Sorry. What I see in your plots are material parameters of an [orthotropic material](https://en.wikipedia.org/wiki/Orthotropic_material). This isn't implemented (but could be defined manually).

**Comment by baneeniram** [comment]
Hello! To embed FDM printing imperfections, such as voids between filaments and layers, into a material model using homogenization, you would need to modify the material's stiffness values based on the average ratio of voids. Typically, this can be achieved by applying a homogenization technique where you account for the porosity (voids) in the material matrix. One common approach is to use a reduced modulus that factors in the percentage of voids present in the material. This reduction can be related to stiffness values by using empirical or analytical methods based on experimental data.

You may also want to explore how specific FDM parameters, such as layer height, print speed, and infill percentage, contribute to void formation. By incorporating these into your model, you can simulate more realistic material properties.

For precise application in local projects, you might consider reaching out to [Qatar 3D print](https://3dversedesign.com/3d-printing-qatar/) services, as they may offer expertise and guidance in tailoring 3D printing processes with specific material and structural properties.
